spec/GraphQLOverHTTP.md

```diff
@@ -188,21 +190,21 @@ Note: both query and mutation operations can be sent over POST requests.
 
 # Response
 
-When a GraphQL server receives a request, it must return a well‐formed response. The server’s
+When a GraphQL server receives a request, it MUST return a well‐formed response. The server’s
 response describes the result of executing the requested operation if successful, and describes
 any errors encountered during the request.
 
 ## Body
 
-If the server's response contains a body it should follow the requirements for [GraphQL response](https://graphql.github.io/graphql-spec/June2018/#sec-Response).
+If the server's response contains a body it SHOULD follow the requirements for [GraphQL response](https://graphql.github.io/graphql-spec/June2018/#sec-Response).
 
-Note: For any non-2XX response, the client should not rely on the body to be in GraphQL format since the source of the response
+Note: For any non-2XX response, the client SHOULD NOT rely on the body to be in GraphQL format since the source of the response
 may not be the GraphQL server but instead some intermediary such as API gateways, firewalls, etc.
 
 ## Status Codes
 
 If the `data` entry in the response has any value other than `null` (when the operation has successfully executed
-without error) then the response should use the 200 (OK) status code.
+without error) then the response SHOULD use the 200 (OK) status code.
 If the operation failed before or during execution, due to a syntax error, missing information, validation error
-or any other reason, then response should use the 4XX or 5XX status codes.
+or any other reason, the response SHOULD use the 4XX or 5XX status codes.
```

Comment:
```suggestion
or any other reason, the response MUST use a 4XX or 5XX status code.
```

Reply:
I'm not sure this is a `MUST`.  The key words to me are `before or during execution`.  If there is a validation error on one of the fields' input arguments, isn't a 2xx response still valid?

Reply:
The current spec has already been updated here, see https://github.com/graphql/graphql-over-http/blob/main/spec/GraphQLOverHTTP.md#status-codes

This PR should be rebased.

Reply:
Agreed

Reply:
>  If there is a validation error on one of the fields' input arguments, isn't a 2xx response still valid?

If it's a GraphQL validation error it occurs before execution so must result in a 4XX/5XX error - the request is invalid and cannot be executed.

If it's an error thrown by a resolver, then it should result in something like:

```graphql
type Mutation {
  myMutation(a: Int): Int
}

mutation ThrowsError {
  myMutation(a: -3)
}
```

```json
{
  "data": {"myMutation": null},
  "errors": [
    ...
  ]
}
```

This should be 200, since "data" is present and non-null.

However in the case that you've defined the mutation such that the field is non-nullable:

```graphql
type Mutation {
  myMutation(a: Int): Int!
}
```

and then the field throws, then the result will be:

```json
{
  "data": null,
  "errors": [
    ...
  ]
}
```

This has been controversial how to handle in the past. There are arguments for both sides. The [above linked document](https://github.com/graphql/graphql-over-http/blob/main/spec/GraphQLOverHTTP.md#status-codes) currently states:

- present and non-null "data" -> 2XX (preferably 200)
- non-2XX -> "data" must be null/not present

It leaves the case of `"data": null` undefined - the text allows for it to be 2XX or non-2XX. Personally I think it should be a 5XX since it's a schema design issue. Take for example this mutation operation:

```graphql
mutation DoesThreeThings {
  createUser(name: "Benjie") { id name }
  orderProduct(id: 27) { deliveryDate }
  myMutation(a: -3)
}
```

If `myMutation` throws (and is non-nullable), the result will be:

```json
{
  "data": null,
  "errors": [
    ...
  ]
}
```

The user cannot know if createUser/orderProduct succeeded (though they can determine that they did not throw errors) due to a schema design fault.

Reply:
@benjie - thanks for the detailed explanation and examples.
If we use your proposal, would we be able to go further and for purposes of HTTP status codes unify no-data and null-data as the same category?
ie:
- present and non-null `data` iff 2xx
- missing or null `data` iff non-2xx

Reply:
I have a vague feeling that’s how express GraphQL works already. I would be happy with this decision, but it should be noted these two situations come from separate phases. I feel like data:null is always a server/schema error, so 5XX.